addon/content/messageHeaderOptions.jsx

```diff
@@ -124,11 +124,44 @@ class MessageHeaderOptions extends React.PureComponent {
   replyAction(msg, event) {
     event.stopPropagation();
     event.preventDefault();
-    this.props.dispatch({
-      ...msg,
-      id: this.props.id,
+
+    const payload = {
       msgUri: this.props.msgUri,
-    });
+      shiftKey: msg.shiftKey,
+    };
+    let action = null;
+    switch (msg.type) {
+      case "reply":
+        action = messageActions.msgReply(payload);
+        break;
+      case "replyAll":
+        action = messageActions.msgReplyAll(payload);
+        break;
+      case "replyList":
+        action = messageActions.msgReplyList(payload);
+        break;
+      case "forward":
+        action = messageActions.msgForward(payload);
+        break;
+      case "editAsNew":
+        action = messageActions.editAsNew(payload);
+        break;
+      case "archive":
+        action = messageActions.msgArchive({ id: this.props.id });
+        break;
+      case "delete":
+        action = messageActions.msgDelete({ id: this.props.id });
+        break;
+      case "classic":
+        action = messageActions.msgOpenClassic(payload);
+        break;
+      case "source":
+        action = messageActions.msgOpenSource(payload);
```

Comment:
We might need to rename some of the types, but I think dropping `msg` here as well would also help simplify - we couldn't replace the whole switch, because of the payloads, but we could treat archive & delete as specials, and then anything else we fallback to `messageActions[type](payload)` maybe?